### PR TITLE
MB-60202 - Filtering deleted documents during FAISS vector index search

### PR DESCRIPTION
**Pull request by metonymic-smokey**
This PR addresses filtering of deleted documents by moving from post-filtering to search time filtering. 

- Passes doc IDs to exclude for each search.
- To account for duplicate vectors: saves every vector with a unique hash so that the selector only excludes the specific vector ID. The current approach of identical vectors having identical hashes would lead to (unnecessarily) excluding vectors if other (identical) vectors in the segment were deleted.

### CODE REVIEW

File: section_faiss_vector_index.go [hunk 1]
```diff
@@ -619,13 +557,19 @@ func (vo *vectorIndexOpaque) process(field index.VectorField, fieldID uint16, do
...
-	var rv int64
+	var rv, sum int64
 	for _, v := range a {
-		rv = int64(math.Float32bits(v)) + (31 * rv)
+		sum += int64(math.Float32bits(v))
 	}
 
+	// Similar to getVectorCode(), this uses the first 32 bits for the vector sum
+	// and the last 32 for a random 32-bit int to ensure identical vectors have
+	// unique hashes.
```
Comment (abhinavdangeti) [inline section_faiss_vector_index.go:571]: Not sure i follow your reasoning here, why do the bitwise OR operation below to obtain a unique hash. Is there some significance in retaining the first 32 bits for the vector sum?

File: section_faiss_vector_index.go [hunk 2]
```diff
@@ -619,13 +557,19 @@ func (vo *vectorIndexOpaque) process(field index.VectorField, fieldID uint16, do
 
 // todo: better hash function?
 // keep the perf aspects in mind with respect to the hash function.
-// random seed based hash golang.
+// Uses a time based seed to avoid 2 identical vectors in different
+// segments having the same hash -> could be an issue when merging those
+// segments
```
Comment (abhinavdangeti) [inline section_faiss_vector_index.go:562]: If you're not going to be de-duplicating vectors from different docIDs - why would there be an issue?
Also - you cannot leave comments like these in here, address them!

File: section_faiss_vector_index.go [hunk 3]
```diff
@@ -657,7 +601,7 @@ type indexContent struct {
 
 type vecInfo struct {
 	vec    []float32
-	docIDs *roaring.Bitmap
+	docIDs uint64
```
Comment (abhinavdangeti) [inline section_faiss_vector_index.go:604]: You changed the type but left the name as it were - change to docID and leave it at uint32.

File: section_faiss_vector_index.go [hunk 4]
```diff
@@ -591,11 +531,9 @@ func (vo *vectorIndexOpaque) process(field index.VectorField, fieldID uint16, do
 		subVecHash := hashCode(subVec)
 		if _, ok := vo.vecIDMap[subVecHash]; !ok {
 			vo.vecIDMap[subVecHash] = vecInfo{
-				docIDs: roaring.NewBitmap(),
+				docIDs: uint64(docNum),
```
Comment (abhinavdangeti) [inline section_faiss_vector_index.go:534]: Why cast a docNum to uint64 here, leave it at uint32. docIDs needs refactoring to docID with your proposal.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: section_faiss_vector_index.go (hunk 2 above)
Comment (metonymic-smokey) [reply section_faiss_vector_index.go:562]: If you're referring to the merge 'issue' in the comment as something to be addressed, it's already addressed when ensuring that no identical vectors have identical hashes. What I meant was that identical vectors having identical hashes was to be avoided during merge.
I think I didn't phrase my comment right.

File: section_faiss_vector_index.go (hunk 1 above)
Comment (metonymic-smokey) [reply section_faiss_vector_index.go:571]: On the off chance that the random number generated is the same for 2 different vectors - this adds an additional layer of differentiation.

File: section_faiss_vector_index.go (hunk 4 above)
Comment (metonymic-smokey) [reply section_faiss_vector_index.go:534]: Did that to avoid casting here - https://github.com/blevesearch/zapx/blob/511fd66ddea9a1301b99f240d42fadb0c3ee7e16/section_faiss_vector_index.go#L495

File: faiss_vector_posting.go [hunk 5]
```diff
@@ -295,22 +295,32 @@ func (sb *SegmentBase) InterpretVectorIndex(field string) (
...
+		for vecID, docID := range vecDocIDMap {
+			if except != nil && except.Contains(docID) {
+				vectorIDsToExclude = append(vectorIDsToExclude, vecID)
+			}
+		}
+
+		var scores []float32
+		var ids []int64
+		var err error
+		scores, ids, err = vecIndex.SearchWithoutIDs(qVector, k, vectorIDsToExclude)
```
Comment (Thejas-bhat) [inline faiss_vector_posting.go:310]: i think `scores, ids, err := vecIndex.SearchWithoutIDs(qVector, k, vectorIDsToExclude)` should suffice over here right?

File: section_faiss_vector_index.go [hunk 6]
```diff
@@ -619,13 +557,19 @@ func (vo *vectorIndexOpaque) process(field index.VectorField, fieldID uint16, do
...
-// random seed based hash golang.
+// Uses a time based seed to avoid 2 identical vectors in different
+// segments having the same hash -> could be an issue when merging those
+// segments
 func hashCode(a []float32) int64 {
-	var rv int64
+	var rv, sum int64
 	for _, v := range a {
-		rv = int64(math.Float32bits(v)) + (31 * rv)
+		sum += int64(math.Float32bits(v))
```
Comment (Thejas-bhat) [inline section_faiss_vector_index.go:566]: could you please explain why the operation (31 * rv) is dropped? like from what i remember this formula i borrowed from java sdk, and from what i understood this factor was something that helped in generating a better hash than sum of values in an array.

Comment (metonymic-smokey) [reply section_faiss_vector_index.go:566]: yes, that's the reason - the main differentiator is now the combination of vector sum + random int. 
Even if 2 different vectors sum to the same number, the random int in the other 32 bits should be the differentiator.

File: cmd/zap/cmd/vector.go [hunk 7]
```diff
@@ -139,32 +138,12 @@ func decodeSection(data []byte, start uint64) (int, int, map[int64][]uint32, *fa
 	numVecs, n := binary.Uvarint(data[pos : pos+binary.MaxVarintLen64])
 	pos += n
 	for i := 0; i < int(numVecs); i++ {
-		vecID, n := binary.Varint(data[pos : pos+binary.MaxVarintLen64])
+		vecID, n := binary.Uvarint(data[pos : pos+binary.MaxVarintLen64])
```
Comment (Thejas-bhat) [inline cmd/zap/cmd/vector.go:141]: this should be binary.Varint() right? like i see that in other parts of the code the vecID is retained as binary.Varint()

**Review by Thejas-bhat — COMMENTED** [review]
just this one comment, and given that you've tested this out, i think the PR makes sense to me.

File: cmd/zap/cmd/vector.go (hunk 7 above)
Comment (metonymic-smokey) [reply cmd/zap/cmd/vector.go:141]: yeah, my bad - accidentally changed this for running some edge case tests.
